test/python/quantum_info/test_quaternions.py

```diff
@@ -92,12 +92,12 @@ def test_mul_by_quat(self):
     def test_mul_by_array(self):
         """Quaternions cannot be multiplied with an array."""
         other_array = np.array([0.1, 0.2, 0.3, 0.4])
-        self.assertRaises(Exception, self.quat_unnormalized.__mul__, other_array)
+        self.assertEqual(NotImplemented, self.quat_unnormalized.__mul__(other_array))
 
     def test_mul_by_scalar(self):
         """Quaternions cannot be multiplied with a scalar."""
         other_scalar = 0.123456789
-        self.assertRaises(Exception, self.quat_unnormalized.__mul__, other_scalar)
+        self.assertEqual(NotImplemented, self.quat_unnormalized.__mul__(other_scalar))
```

Comment:
This is pre-existing, stemming from the previous implementation of these magic methods being incorrect, but in general we don't want to call `__mul__` directly - `NotImplemented` is just a sentinel value that's properly interpreted by Python.  The normal forms of these tests would be
```python
with self.assertRaises(TypeError):
    _ = self.quat_unnormalized * other_array
```
and similar for the other one.

This actually means that Python will also try the `__rmul__` methods on `other_array` and `other_scalar` as well before raising the `TypeError`, in case either of those classes is something that can explicitly handle multiplication by `Quaternion`, though in this case, both will fail.  The Numpy array test will actually _start_ to succeed (Numpy will attempt to broadcast the multiplication), but then it'll fail the internal scalar multiplication.

Reply:
i see. yeah, i was thinking this was more of a unit test vs integration test. updates on the way!